### ISSUE TITLE
Inconsistencies between default and custom MAPE 

### ISSUE THREAD

**Issue by sweronika** [catboost/catboost#1947]
catboost version: 1.0.3

Hi.
I wanted to implement custom loss function (SMAPE) in catboost algorithm. 
I decided to start with implementing custom MAPE in order to compare my results with the ones obtained with default MAPE loss that is already implemented. 
I followed the example from catboost website for implementing custom loss and I used the exact same derivative for MAPE that is implemented in error_functions.sh script. However I keep geting different results. Because of that I'm not even trying to implement SMAPE as even regular custom MAPE does not work for me.
Could you please explain where do the inconsistencies come from?

```
### Custom MAPE according to equation from catboost source code

class MAPEObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            
            if (targets[index] - approxes[index]) > 0:
                der1 = 1/max(1, abs(targets[index]))
            else:
                der1 = -1/max(1, abs(targets[index]))
                
            der2 = 0

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result

# Initialize CatBoostRegressor
model = CatBoostRegressor(loss_function=MAPEObjective(),
                          eval_metric='MAPE',
                          cat_features=categorical_features_indices,
                          iterations=2,
                          learning_rate=1,
                          depth=2)

# Fit model
model.fit(X_train, y_train, eval_set=(X_validation, y_validation))
# Get predictions
preds = model.predict(X_validation)
```

The outcomes are:
```
0:	learn: 0.6613650	test: 0.6393375	best: 0.6393375 (0)	total: 352ms	remaining: 352ms
1:	learn: 0.8055784	test: 0.7312165	best: 0.6393375 (0)	total: 355ms	remaining: 0us

bestTest = 0.6393374803
bestIteration = 0
```
And when running the algorithm with MAPE that is already implemented:
```
# Initialize CatBoostRegressor
model = CatBoostRegressor(loss_function='MAPE',
                          eval_metric='MAPE',
                          cat_features=categorical_features_indices,
                          iterations=2,
                          learning_rate=1,
                          depth=2)

# Fit model
model.fit(X_train, y_train, eval_set=(X_validation, y_validation))
# Get predictions
preds = model.predict(X_validation)
```

The outcomes are quite different:
```
0:	learn: 0.5405275	test: 0.5179807	best: 0.5179807 (0)	total: 4.5ms	remaining: 4.5ms
1:	learn: 0.5170359	test: 0.4831936	best: 0.4831936 (1)	total: 7.43ms	remaining: 0us

bestTest = 0.483193642
bestIteration = 1
```
It was run for test purposes on titanic dataset. 
While running catboost with both functions on my real-case dataset the differences were even more significant.
I can share the full code if you need it. 
I would be grateful for some help.

**Comment by ek-ak** [comment]
Hello! In CatBoost MAPE is used with defaults
```
leaf_estimation_method='Exact',
boost_from_average=True
```
which is not allowed for user-defined metrics  now =(
With options 
```
leaf_estimation_method='Gradient',
boost_from_average=False
```
your MAPE realization is equivalent to CatBoost MAPE.
You can check model options with method `get_all_params()`
`jsondiff.diff(model.get_all_params(), model2.get_all_params())`

**Comment by sweronika** [comment]
Thank you for quick response, that helped a lot!
I also found out that one more hyperparameter - leaf_estimation_backtracking is not completely supported with custom loss function. I trained the models with mentioned hyperparameters set as you suggested, however there are still some inconsistencies. All hyperparameters seems to be the same now for both models. Do you have any hints on why there are still those differences?

```
model_default = CatBoostRegressor(loss_function='MAPE',
                          eval_metric='MAPE',
                          cat_features=categorical_features_indices,
                          leaf_estimation_method='Gradient',
                          boost_from_average=False,
                          leaf_estimation_backtracking='No',
                          iterations=2,
                          learning_rate=1,
                          depth=2)
# Fit model
model_default.fit(X_train, y_train, eval_set=(X_validation, y_validation))
# Get predictions
preds = model_default.predict(X_validation)

0:	learn: 0.9778114	test: 0.9868911	best: 0.9868911 (0)	total: 4.54ms	remaining: 4.54ms
1:	learn: 0.9717578	test: 0.9809798	best: 0.9809798 (1)	total: 7.79ms	remaining: 0us

bestTest = 0.9809797916
bestIteration = 1
```

```
model_custom = CatBoostRegressor(loss_function=MAPEObjective(),
                          eval_metric='MAPE',
                          cat_features=categorical_features_indices,
                          leaf_estimation_method='Gradient',
                          boost_from_average=False,
                          leaf_estimation_backtracking='No',
                          iterations=2,
                          learning_rate=1,
                          depth=2)

# Fit model
model_custom.fit(X_train, y_train, eval_set=(X_validation, y_validation))
# Get predictions
preds = model_custom.predict(X_validation)

0:	learn: 0.9881736	test: 0.9970757	best: 0.9970757 (0)	total: 23.9ms	remaining: 23.9ms
1:	learn: 0.9951439	test: 1.0033814	best: 0.9970757 (0)	total: 46.8ms	remaining: 0us

bestTest = 0.9970757026
bestIteration = 0
```

**Comment by ek-ak** [comment]
Hello! Sorry for not answering for a long time. Could you provide reproducible example?
This code gives equal results:
```
import pandas as pd
import os
import numpy as np
np.set_printoptions(precision=4)
import catboost
from catboost import *
from catboost import datasets
from sklearn.model_selection import train_test_split


(train_df, test_df) = catboost.datasets.amazon()
y = train_df.ACTION
X = train_df.drop('ACTION', axis=1)
cat_features = list(range(0, X.shape[1]))
X_train, X_validation, y_train, y_validation = train_test_split(X, y, train_size=0.8, random_state=1234)


class MAPEObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            if (targets[index] - approxes[index]) > 0:
                der1 = 1/max(1, abs(targets[index]))
            else:
                der1 = -1/max(1, abs(targets[index]))
                
            der2 = 0

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result



parameters = dict(
    eval_metric='MAPE',
    cat_features=cat_features,
    iterations=4,
    learning_rate=1,
    leaf_estimation_method='Gradient',
    boost_from_average=False,
    leaf_estimation_backtracking='No',
    depth=2,
    loss_function='MAPE'
)
model = CatBoostRegressor(**parameters)
model.fit(X_train, y_train, eval_set=(X_validation, y_validation))
print(model.predict(X_validation))

parameters['loss_function'] = MAPEObjective()
model2 = CatBoostRegressor(**parameters)
model2.fit(X_train, y_train, eval_set=(X_validation, y_validation))
print(model2.predict(X_validation))

0:	learn: 0.1587948	test: 0.1592788	best: 0.1592788 (0)	total: 5.53ms	remaining: 16.6ms
1:	learn: 0.8259745	test: 0.8302054	best: 0.1592788 (0)	total: 11.2ms	remaining: 11.2ms
2:	learn: 0.2594193	test: 0.2564103	best: 0.1592788 (0)	total: 15.1ms	remaining: 5.02ms
3:	learn: 0.7105553	test: 0.7146898	best: 0.1592788 (0)	total: 18.7ms	remaining: 0us

bestTest = 0.1592787708
bestIteration = 0

Shrink model to first 1 iterations.
[0.889 0.889 0.889 ... 0.889 0.889 0.889]
0:	learn: 0.1587948	test: 0.1592788	best: 0.1592788 (0)	total: 361ms	remaining: 1.08s
1:	learn: 0.8259745	test: 0.8302054	best: 0.1592788 (0)	total: 400ms	remaining: 400ms
2:	learn: 0.2594193	test: 0.2564103	best: 0.1592788 (0)	total: 434ms	remaining: 145ms
3:	learn: 0.7105553	test: 0.7146898	best: 0.1592788 (0)	total: 474ms	remaining: 0us

bestTest = 0.1592787708
bestIteration = 0

Shrink model to first 1 iterations.
[0.889 0.889 0.889 ... 0.889 0.889 0.889]
```

**Comment by kizill** [comment]
I hope, full sample, provided by @ek-ak should resolve this issue, as results are consistent. If you still have original problem - please, reopen that issue!
